staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go

```diff
@@ -110,16 +116,27 @@ func (c *APIServiceRegistrationController) getDestinationHost(apiService *apireg
 	if err != nil {
 		return destinationHost
 	}
-	switch {
+	switch service.Spec.Type {
 	// use IP from a clusterIP for these service types
-	case service.Spec.Type == v1.ServiceTypeClusterIP,
-		service.Spec.Type == v1.ServiceTypeNodePort,
-		service.Spec.Type == v1.ServiceTypeLoadBalancer:
-		return service.Spec.ClusterIP
+	case v1.ServiceTypeClusterIP, v1.ServiceTypeNodePort, v1.ServiceTypeLoadBalancer:
+		if service.Spec.ClusterIP == v1.ClusterIPNone {
+			return ""
+		}
+		destinationHost = service.Spec.ClusterIP
+	case v1.ServiceTypeExternalName:
+		destinationHost = service.Spec.ExternalName
+		if len(destinationHost) == 0 {
+			return ""
+		}
+	}
+
+	port := int32(443)
+	if len(service.Spec.Ports) > 0 {
+		port = service.Spec.Ports[0].Port
```

Comment:
I assumed that there is an order. If not, this does not make sense, totally agree. Actually, I do not want too many heuristic. I would prefer adding an explicit port name to the apiservice spec and only allow to leave that empty if there is exactly one port in a service. Wdyt?